### PR TITLE
refactor: move to PevmChain::get_tx_env

### PR DESCRIPTION
**Pull request by kien-rise**
Resolves https://github.com/risechain/pevm/pull/233#discussion_r1745009221

### CODE REVIEW

File: src/chain.rs [hunk 1]
```diff
@@ -61,9 +61,12 @@ pub trait PevmChain: Debug {
     fn calculate_receipt_root(
         &self,
         spec_id: SpecId,
-        txs: &BlockTransactions<Transaction>,
+        txs: &BlockTransactions<Self::Transaction>,
         tx_results: &[PevmTxExecutionResult],
     ) -> B256;
+
+    /// Get [TxEnv]
+    fn get_tx_env(&self, tx: Self::Transaction) -> Result<TxEnv, Self::TxEnvError>;
```
Comment (hai-rise) [inline src/chain.rs:69]: Minor but let's put this where `get_gas_price` was to flow better.

File: src/chain/ethereum.rs [hunk 2]
```diff
@@ -55,9 +62,28 @@ pub enum EthereumGasPriceError {
...
+            .ok_or(EthereumGasPriceError::MissingGasPrice),
+        TxType::Eip1559 | TxType::Eip4844 | TxType::Eip7702 => tx
+            .max_fee_per_gas
+            .map(U256::from)
+            .ok_or(EthereumGasPriceError::MissingMaxFeePerGas),
+    }
+}
+
 impl PevmChain for PevmEthereum {
+    type Transaction = Transaction;
```
Comment (hai-rise) [inline src/chain/ethereum.rs:84]: Rare place where we should add `::` to be honest.

File: src/chain/ethereum.rs [hunk 3]
```diff
@@ -193,4 +201,29 @@ impl PevmChain for PevmEthereum {
         }
         hash_builder.root()
     }
+
+    /// Get the REVM tx envs of an Alloy block.
+    // https://github.com/paradigmxyz/reth/blob/280aaaedc4699c14a5b6e88f25d929fe22642fa3/crates/primitives/src/revm/env.rs#L234-L339
+    // https://github.com/paradigmxyz/reth/blob/280aaaedc4699c14a5b6e88f25d929fe22642fa3/crates/primitives/src/alloy_compat.rs#L112-L233
+    // TODO: Properly test this.
+    fn get_tx_env(&self, tx: Self::Transaction) -> Result<TxEnv, EthereumTxEnvError> {
```
Comment (hai-rise) [inline src/chain/ethereum.rs:209]: Put this where `get_gas_price` was to flow smoother.

File: src/chain/ethereum.rs [hunk 4]
```diff
@@ -193,4 +201,29 @@ impl PevmChain for PevmEthereum {
...
+        Ok(TxEnv {
+            caller: tx.from,
+            gas_limit: tx
+                .gas
+                .try_into()
+                .map_err(|_| EthereumTxEnvError::OverflowedGasLimit)?,
+            gas_price: get_ethereum_gas_price(&tx).map_err(EthereumTxEnvError::GasPriceError)?,
```
Comment (hai-rise) [inline src/chain/ethereum.rs:216]: This unnecessarily overwrites `EthereumGasPriceError`?  `get_ethereum_gas_price` can just return `EthereumTxEnvError` and/or consider inlining it here.

File: src/chain/ethereum.rs [hunk 5]
```diff
@@ -55,9 +62,28 @@ pub enum EthereumGasPriceError {
...
+            .map(U256::from)
+            .ok_or(EthereumGasPriceError::MissingMaxFeePerGas),
+    }
+}
+
 impl PevmChain for PevmEthereum {
+    type Transaction = Transaction;
     type BlockSpecError = EthereumBlockSpecError;
-    type GasPriceError = EthereumGasPriceError;
+    type TxEnvError = EthereumTxEnvError;
```
Comment (hai-rise) [inline src/chain/ethereum.rs:86]: `TransactionParsingError` sounded better to be honest.

File: src/chain/ethereum.rs (hunk 2 above)
Comment (kien-rise) [reply src/chain/ethereum.rs]: Can you check again? I explicit write `alloy_rpc_types::Transaction` in this place and some more places.

File: src/chain/ethereum.rs [hunk 6]
```diff
@@ -20,6 +20,19 @@ use crate::{
     mv_memory::MvMemory, BuildIdentityHasher, MemoryLocation, PevmTxExecutionResult, TxIdx,
 };
 
+/// Represents errors that can occur when parsing transactions
+#[derive(Debug, Clone, PartialEq)]
+pub enum EthereumTransactionParsingError {
```
Comment (hai-rise) [inline src/chain/ethereum.rs:25]: Very minor but this should be placed where `EthereumGasPriceError` was.

**Review by hai-rise — APPROVED** [review]

**Pull request merged**